src/ParticleContainer.cpp

```diff
@@ -0,0 +1,84 @@
+//
+// Created by Celik Yucel on 23.10.24.
+//
+
+    #include "ParticleContainer.h"
+
+    #include <algorithm>
+
+    std::vector<Particle>::iterator ParticleContainer::begin() {
+        return particles.begin();
+    }
+    std::vector<Particle>::iterator ParticleContainer::end() {
+        return particles.end();
+    }
+
+    std::vector<std::pair<std::reference_wrapper<Particle>, std::reference_wrapper<Particle>>>::iterator ParticleContainer::begin_pairs() {
+        return particlePairs.begin();
+    }
+
+    std::vector<std::pair<std::reference_wrapper<Particle>, std::reference_wrapper<Particle> > >::iterator ParticleContainer::end_pairs() {
+        return particlePairs.end();
+    }
+
+
+
+    void ParticleContainer::addParticle(const Particle &particle) {
+        particles.push_back(particle);
+    }
+
+    void ParticleContainer::removeParticle(const Particle &particle) {
+        particles.erase(
+            std::remove_if(particles.begin(), particles.end(), [&particle](const Particle &p) {
+                    if (&p == &particle) {
+                        return true;
+                    }
+                    return false;
+            }),
+            particles.end());
+        reinitializePairs();
+    }
+
+    bool ParticleContainer::pairExists(const Particle &particle1, const Particle &particle2) const {
+        for (const auto &p: particlePairs) {
+            if (((&p.first.get() == &particle1) && (&p.second.get() == &particle2)) ||
+                ((&p.first.get() == &particle2) && (&p.second.get() == &particle1))) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void ParticleContainer::initializePairsVector() {
+        for (auto iterator1 = particles.begin(); iterator1 != particles.end(); ++iterator1) {
+            for (auto iterator2 = iterator1 + 1; iterator2 != particles.end(); ++iterator2) {
+                if (!pairExists(*iterator1, *iterator2)) {
+                    particlePairs.push_back({*iterator1,*iterator2});
+                }
+            }
+        }
+    }
```

Comment:
This is a nice approach and useful for your present case. However, this might lead to a lot of overhead if you need to add (and even slight overhead in removing particles) more dynamically. Because then you will need to reinitialize a lot and it is O(n^2). But leave it as it is, as you will implement a new container (LinkedCells) soon and use that for last 3 worksheets.

Reply:
We thought iterating through the particle pairs would be useful for our calculations, but to iterate like this we have to go through every particle twice in any case. We are not sure if there is a better way to do this/ get a better run time. We would be thankful if you could push us in the right direction.

Reply:
I would suggest instead of creating pairs, just create .begin() and .end() iterators that return first and last pair, and provide a appropriate update (++) mechanism. so for eg. with particles p1, p2, p3, p4, begin can return (p1,p2), update can change to, (p1,p3), and then (p1,p4), then (p2,p3), and so on. This would be one solution but you can think of you own.

However, I would suggest to leave it for now and focus on finishing the worksheet 2 tasks, come back to it afterwards if you have time :)